src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1085,47 +1102,139 @@ private static final class SimpleStringBuilderStrategy {
+                                .withField(LENGTH, CD_int, FINAL_STABLE_FIELD)
+                                .withField(CODER, CD_byte, FINAL_STABLE_FIELD)
+                                .withField(CONSTANTS, CD_Array_String, FINAL_STABLE_FIELD)
+                                .withMethodBody("<init>",
+                                        ARRAY_STRING_TO_VOID,
+                                        ClassFile.ACC_PRIVATE,
+                                        generateConstructor(concatClass, constants.length)
+                                )
                                 .withMethodBody(METHOD_NAME,
-                                        ConstantUtils.methodTypeDesc(erasedArgs),
-                                        ClassFile.ACC_FINAL | ClassFile.ACC_PRIVATE | ClassFile.ACC_STATIC,
-                                        generateMethod(constants, erasedArgs));
+                                        ConstantUtils.methodTypeDesc(concatArgs),
+                                        ClassFile.ACC_FINAL | ClassFile.ACC_PRIVATE,
+                                        generateConcatMethod(concatClass, concatArgs));
                     }});
             try {
                 var hiddenClass = lookup.makeHiddenClassDefiner(className, classBytes, Set.of(), DUMPER)
                         .defineClass(true, null);
-                return lookup.findStatic(hiddenClass, METHOD_NAME, erasedArgs);
-            } catch (Exception e) {
+
+                MethodHandle constructorHandle = lookup.findConstructor(hiddenClass, MethodType.methodType(void.class, String[].class));
+                var instance = hiddenClass.cast(constructorHandle.invoke(constants));
+                MethodHandle handle = lookup.findVirtual(hiddenClass, METHOD_NAME, concatArgs);
+                CACHE.put(concatArgs, new SoftReference<>(new MethodHandlePair(constructorHandle, handle)));
+                return handle.bindTo(instance);
+            } catch (Throwable e) {
                 throw new StringConcatException("Exception while spinning the class", e);
             }
         }
 
+        private static Consumer<CodeBuilder> generateConstructor(ClassDesc concatClass, int numConstants) {
+            return new Consumer<CodeBuilder>() {
+                @Override
+                public void accept(CodeBuilder cb) {
+                    cb.aload(0);
+                    cb.invokespecial(CD_Object, INIT_NAME, MTD_void, false);
+                    cb.aload(0);
+                    cb.aload(1);
+                    cb.putfield(concatClass, CONSTANTS, CD_Array_String);
+                    cb.iconst_0();
+                    cb.istore(2);
+                    cb.iconst_0();
+                    cb.istore(3);
+                    cb.iconst_0();
+                    cb.istore(4);
+                    Label loopStart = cb.newLabel();
+                    Label loopEnd = cb.newLabel();
```

Comment:
These can be handled by BlockCodeBuilder:
```java
cb.block(bcb -> {
  // ...
  bcb.if_icmpge(bcb.endLabel());
  // ...
  bcb.goto_(bcb.startLabel());
);
});
```
You can also pre-compute the stack maps at the block code builder start (uninitializedThis, String[], int length, int coder, int i; empty stack) and end (same locals + stack) labels to save like 15% of the stack map generation overhead.